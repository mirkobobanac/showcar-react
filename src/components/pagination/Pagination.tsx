import * as React from 'react'
import { Dispatch } from 'redux'

type Props = {
  /**
   * number of items per page
   */
  pageSize: number

  /**
   * current selected page
   */
  currentPage: number

  /**
   * total number of items
   */
  totalItems: number

  /**
   * Text to use for the prev/next buttons
   */
  messages: {
    next: string
    previous: string
  }

  /**
   * Callback to trigger when a page is selected
   */
  onPageSelection: (page: number, url: string) => void

  /**
   * url template where pagination will be injected. Must have placeholder `&page={page}`
   */
  urlTemplate: string

  /**
   * Due to underlying showcar-ui, there is a chance the pagination fails to build. This callback will receive the error in such cases
   */
  onCreationError?: (e: Error) => void

  /**
   * Whether to scroll to the top of the page upon page change. Defaults to false
   */
  autoScrollToTop?: boolean
}

/**
 * Pagination component
 */
class Pagination extends React.Component<Props> {
  private paginationElement!: HTMLElement // assigned on render

  public render() {
    return (
      <div className="cl-pagination">
        <ul
          className="sc-pagination"
          ref={r => r && (this.paginationElement = r)}
          data-previous-text={this.props.messages.previous}
          data-next-text={this.props.messages.next}
        />
      </div>
    )
  }

  public componentDidMount() {
    this.updatePager()
  }

  /**
   * TODO: Refactor to not use componentWillReceiveProps
   */
  public UNSAFE_componentWillReceiveProps(nextProps: Readonly<Props>, nextContext: any) {
    if (this.props !== nextProps) {
      this.updatePager()
    }
  }

  public componentDidUpdate(prevProps: Props) {
    if (
      prevProps.pageSize !== this.props.pageSize ||
      prevProps.currentPage !== this.props.currentPage ||
      prevProps.totalItems !== this.props.totalItems ||
      prevProps.urlTemplate !== this.props.urlTemplate ||
      prevProps.messages !== this.props.messages ||
      prevProps.urlTemplate !== this.props.urlTemplate ||
      prevProps.autoScrollToTop !== this.props.autoScrollToTop
    ) {
      this.updatePager()
    }

    if (prevProps.currentPage !== this.props.currentPage && this.props.autoScrollToTop) {
      this.focusTop()
    }
  }

  private focusTop() {
    window.scrollTo(0, 0)
  }

  private updatePager() {
    if (this.paginationElement && window.Pager) {
      try {
        // tslint:disable-next-line:no-unused-expression
        new window.Pager(
          this.paginationElement,
          this.props.pageSize,
          this.props.currentPage,
          this.props.totalItems,
          this.props.urlTemplate,
          false
        )
      } catch (e) {
        this.props.onCreationError && this.props.onCreationError(e)
      }
    }

    Array.from(this.paginationElement.getElementsByTagName('a'))
      .map(elem => ({
        elem,
        url: elem.getAttribute('href')
      }))
      .map(elementAndUrl => ({
        elem: elementAndUrl.elem,
        page: elementAndUrl.url ? extractIntParamNumberFromUrl(elementAndUrl.url)('page') : null
      }))
      .forEach(data => {
        data.elem.onclick = (ev: MouseEvent) => {
          ev.preventDefault()
          // console.log(data)
          if (data.page !== null && data.page !== 0 && !isNaN(data.page)) {
            // console.log('triggering page selection', this.props.onPageSelection)
            this.props.onPageSelection(data.page, data.elem.getAttribute('href')!)
          }
        }
      })
  }
}

export const extractIntParamNumberFromUrl = (url: string) => (param: string): number | null => {
  const parts = url.split(`&${param}=`).length > 1 ? url.split(`&${param}=`) : url.split(`?${param}=`)

  if (parts.length > 1) {
    const n = Number(parts[1].split('&')[0])
    return !isNaN(n) ? n : null
  } else {
    return null
  }
}

export default Pagination
