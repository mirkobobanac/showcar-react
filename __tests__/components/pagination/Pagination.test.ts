import initStoryshots from '@storybook/addon-storyshots'
import Pagination, { extractIntParamNumberFromUrl } from '../../../src/components/pagination/Pagination'

initStoryshots({
  storyKindRegex: /\/Pagination$/
})
test('Extracts int parameter from a url', () => {
  expect(extractIntParamNumberFromUrl('/?foo=2')('foo')).toBe(2)
  expect(extractIntParamNumberFromUrl('/?&foo=2&o=1')('foo')).toBe(2)
  expect(extractIntParamNumberFromUrl('/?o=1')('foo')).toBe(null)
  expect(extractIntParamNumberFromUrl('?foo=bar&o=1')('foo')).toBe(null)
  expect(extractIntParamNumberFromUrl('')('foo')).toBe(null)
})
