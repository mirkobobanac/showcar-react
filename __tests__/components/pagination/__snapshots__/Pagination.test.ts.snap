// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Core|Organisms/Pagination Navigation elements 1`] = `
<div>
  <div
    className="cl-pagination"
  >
    <ul
      className="sc-pagination"
      data-next-text="next"
      data-previous-text="prev"
    />
  </div>
</div>
`;

exports[`Storyshots Core|Organisms/Pagination autoscroll to top 1`] = `
Array [
  "Scroll to bottom to interact w/ pagination...",
  <div
    style={
      Object {
        "height": "1600px",
      }
    }
  />,
  <span>
    Choose a page to test autoScroll behaviour...
  </span>,
  <br />,
  <br />,
  <div
    className="cl-pagination"
  >
    <ul
      className="sc-pagination"
      data-next-text="next"
      data-previous-text="prev"
    />
  </div>,
]
`;

exports[`Storyshots Core|Organisms/Pagination usage 1`] = `
<div>
  <div
    style={Object {}}
  >
    <div
      className="cl-pagination"
    >
      <ul
        className="sc-pagination"
        data-next-text="next"
        data-previous-text="prev"
      />
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "WebkitFontSmoothing": "antialiased",
          "backgroundColor": "#fff",
          "border": "1px solid #eee",
          "borderRadius": "2px",
          "boxShadow": "0px 2px 3px rgba(0, 0, 0, 0.05)",
          "color": "#444",
          "fontFamily": "-apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", \\"Arial\\", sans-serif",
          "fontSize": "15px",
          "fontWeight": 300,
          "lineHeight": 1.45,
          "marginBottom": "20px",
          "marginTop": "20px",
          "padding": "20px 40px 40px",
        }
      }
    >
      
      <div>
        <h1
          style={
            Object {
              "borderBottom": "1px solid #EEE",
              "fontSize": "25px",
              "margin": "20px 0 0 0",
              "padding": "0 0 5px 0",
            }
          }
        >
          Prop Types
        </h1>
        <div>
          <h2
            style={
              Object {
                "margin": "20px 0 0 0",
              }
            }
          >
            "
            Pagination
            " Component
          </h2>
          <small>
            No propTypes defined!
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
